pkg/controller.v2/controller_pod_test.go

```diff
@@ -0,0 +1,125 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	tfv1alpha2 "github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1alpha2"
+	tfjobclientset "github.com/kubeflow/tf-operator/pkg/client/clientset/versioned"
+)
+
+func newBasePod(name string, tfJob *tfv1alpha2.TFJob, t *testing.T) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Labels:          genLabels(getKey(tfJob, t)),
+			Namespace:       tfJob.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(tfJob, controllerKind)},
+		},
+	}
+}
+
+func newPod(tfJob *tfv1alpha2.TFJob, typ string, index int, t *testing.T) *v1.Pod {
+	pod := newBasePod(fmt.Sprintf("%s-%d", typ, index), tfJob, t)
+	pod.Labels[tfReplicaTypeLabel] = typ
+	pod.Labels[tfReplicaIndexLabel] = fmt.Sprintf("%d", index)
+	return pod
+}
+
+// create count pods with the given phase for the given tfJob
+func newPodList(count int32, status v1.PodPhase, tfJob *tfv1alpha2.TFJob, typ string, start int32, t *testing.T) []*v1.Pod {
+	pods := []*v1.Pod{}
+	for i := int32(0); i < count; i++ {
+		newPod := newPod(tfJob, typ, int(start+i), t)
+		newPod.Status = v1.PodStatus{Phase: status}
+		pods = append(pods, newPod)
+	}
+	return pods
+}
+
+func setPodsStatuses(podIndexer cache.Indexer, tfJob *tfv1alpha2.TFJob, typ string, pendingPods, activePods, succeededPods, failedPods int32, t *testing.T) {
+	var index int32
+	for _, pod := range newPodList(pendingPods, v1.PodPending, tfJob, typ, index, t) {
+		podIndexer.Add(pod)
+	}
+	index += pendingPods
+	for _, pod := range newPodList(activePods, v1.PodRunning, tfJob, typ, index, t) {
+		podIndexer.Add(pod)
+	}
+	index += activePods
+	for _, pod := range newPodList(succeededPods, v1.PodSucceeded, tfJob, typ, index, t) {
+		podIndexer.Add(pod)
+	}
+	index += succeededPods
+	for _, pod := range newPodList(failedPods, v1.PodFailed, tfJob, typ, index, t) {
+		podIndexer.Add(pod)
+	}
+}
+
+func TestAddPod(t *testing.T) {
+	// Prepare the clientset and controller for the test.
+	kubeClientSet := kubeclientset.NewForConfigOrDie(&rest.Config{
+		Host: "",
+		ContentConfig: rest.ContentConfig{
+			GroupVersion: &v1.SchemeGroupVersion,
+		},
+	},
+	)
+	tfJobClientSet := tfjobclientset.NewForConfigOrDie(&rest.Config{
+		Host: "",
+		ContentConfig: rest.ContentConfig{
+			GroupVersion: &tfv1alpha2.SchemeGroupVersion,
+		},
+	},
+	)
+	controller, _, tfJobInformerFactory := newTFJobControllerFromClient(kubeClientSet, tfJobClientSet, NoResyncPeriodFunc)
+	controller.tfJobListerSynced = alwaysReady
+	controller.podListerSynced = alwaysReady
+	controller.serviceListerSynced = alwaysReady
+	tfJobIndexer := tfJobInformerFactory.Kubeflow().V1alpha2().TFJobs().Informer().GetIndexer()
+
+	stopCh := make(chan struct{})
+	run := func(<-chan struct{}) {
+		controller.Run(threadCount, stopCh)
+	}
+	go run(stopCh)
+
+	var key string
```

Comment:
Yeah I think so. I will open an issue to keep track. I think it is better not to import more enhancements in this PR since it is copied from upstream/v1alpha2 branch and I think we could record the potential issues and fix it after it is merged.

WDYT?

Reply:
Yes, opening an issue and tracking would be ok.